torch/_dynamo/guards.py

```diff
@@ -311,6 +312,23 @@ def uninteresting_files():
     return {inspect.getfile(m) for m in mods}
 
 
+def check_overlapping(overlapping, non_overlapping):
+    from torch._functorch._aot_autograd.input_output_analysis import (
+        _tensors_definitely_do_not_overlap,
+    )
+
+    tensors = overlapping + non_overlapping
+    overlapping_indices = set()
+
+    for i in range(len(tensors)):
+        for j in range(i):
+            if not _tensors_definitely_do_not_overlap(tensors[i], tensors[j]):
+                overlapping_indices.add(i)
+                overlapping_indices.add(j)
+
+    return all(i in overlapping_indices for i in range(len(overlapping)))
```

Comment:
this guard as written will "pass" as long as the overlapping indices we discover at runtime are a superset of the overlapping indices we discovered at compile time.

That seems... wrong? Say we had:

```
def f(a, b, c, d):
    a.mul_(2)
    return a + b + c +d
    
# call 1: (a, b) overlap
# call 2: (a, b, c) overlap
```

Then the first graph will have generated a graph with `a_updated` and `b_updated` (since they were both mutated), but for the second call we would need a graph where `a/b/c` were all updated.

So I guess we have two options:

(1) make the overlapping guard an exact match: for a given alias set, we require the exact list of input indices to be overlapping / not overlapping 

(2) make the overlapping guard overly conservative: in particular, if (a,b,c) were "maybe overlapping" on the first invocation, then it would technically be safe to re-use that graph in a case where only (a, b) are overlapping.

Reply:
Also playing around a little locally, I found an example where we are silently incorrect today (even when there are no dynamic shapes) - I'm curious if this PR helps:
```
import torch


def f(x, y):
    x.mul_(2)
    return x + y


x1 = torch.ones(4)
y1 = torch.ones(4)

f_compiled = torch.compile(f, backend="aot_eager")
out1 = f_compiled(x1, y1)

base = torch.ones(8)
x2 = base[:4]
y2 = base[2:6]
# None of the inputs alias
out2 = f_compiled(x2, y2)

base_ref = torch.ones(8)
x2_ref = base_ref[:4]
y2_ref = base_ref[2:6]
# This should re-compile, since inputs now alias!
# But it does not
out2_ref = f(x2_ref, y2_ref)

print(out2_ref)
print(out2)
```

Reply:
> this guard as written will "pass" as long as the overlapping indices we discover at runtime are a superset of the overlapping indices we discovered at compile time.

You are right. That was definitely not the intention here. Although (2) may be profitable for long compile times, I think that for this PR we should go for (1). It's easier and sound.

Reply:
> Also playing around a little locally, I found an example where we are silently incorrect today

I don't think this PR solves that issue. The problem with that is: since the first compilation doesn't have alias mutation, no tensor overlapping guards is installed.

Reply:
makes sense. filed a separate issue https://github.com/pytorch/pytorch/issues/139789